YAROD - Yet Another Round Of Docs

mavam opened the pull request:
This is another PR with a potpourri of updates.

dit7ya commented:
Some docs issue that I noticed. Will write more later.

1. https://docs.tenzir.com/next/operators/transformations/drop -> is the word `dual` familiar to our target audience?
2. https://docs.tenzir.com/next/operators/sources/load -> typo in the link to connectors
3. https://docs.tenzir.com/next/operators/sources/read -> Not clear, if it is a shortform, add an example here that does not use the word `from` + synopsis.
4. https://docs.tenzir.com/next/operators/sources/shell -> Add synopsis and example nonetheless.
5. https://docs.tenzir.com/next/operators/transformations/drop -> 
> A list of extractors identifying fields to remove. 

what is the list syntax ? (if it is comma write `list of comma separated extractors)

mavam commented:
> Some docs issue that I noticed. Will write more later.

Thanks! Keep 'em coming.

> 1. https://docs.tenzir.com/next/operators/transformations/drop -> is the word `dual` familiar to our target audience?

Perhaps not. I moved it from the very top down into the description so that it's less prominent.

> 2. https://docs.tenzir.com/next/operators/sources/load -> typo in the link to connectors

Fixed.

> 3. https://docs.tenzir.com/next/operators/sources/read -> Not clear, if it is a shortform, add an example here that does not use the word `from` + synopsis.

I've added examples.

> 4. https://docs.tenzir.com/next/operators/sources/shell -> Add synopsis and example nonetheless.

Done.

> 5. https://docs.tenzir.com/next/operators/transformations/drop ->
> 
> > A list of extractors identifying fields to remove.
> 
> what is the list syntax ? (if it is comma write `list of comma separated extractors)

Clarified this per your suggestion.

mavam commented:
> Some docs issue that I noticed. Will write more later.

Thanks! Keep 'em coming.

> 1. https://docs.tenzir.com/next/operators/transformations/drop -> is the word `dual` familiar to our target audience?

Perhaps not. I moved it from the very top down into the description so that it's less prominent.

> 2. https://docs.tenzir.com/next/operators/sources/load -> typo in the link to connectors

Fixed.

> 3. https://docs.tenzir.com/next/operators/sources/read -> Not clear, if it is a shortform, add an example here that does not use the word `from` + synopsis.

I've added examples.

> 4. https://docs.tenzir.com/next/operators/sources/shell -> Add synopsis and example nonetheless.

Done.

> 5. https://docs.tenzir.com/next/operators/transformations/drop ->
> 
> > A list of extractors identifying fields to remove.
> 
> what is the list syntax ? (if it is comma write `list of comma separated extractors)

Clarified this per your suggestion.

dominiklohmann commented on web/docs/get-started.md:
```diff
@@ -205,6 +205,14 @@ storage.
+:::caution Unsafe Pipelines
+Some operators are inherently unsafe because of their side effects, e.g.,
+reading a file, acquiring packets from the network, or using a third-party
+library. We do not allow these operators by default in managed pipelines. You
```
I don't think we ever use the term "managed pipeline" outside of code.

mavam replied:
I'm open to suggestions.

We need to differentiate them from pipelines created with the `tenzir` binary. Something along the lines of "pipelines managed by a node" or "pipelines created through the API".

dominiklohmann replied:
The restriction applies to all _local_ operators that are run remotely instead. This includes:
- All _local_ operators that are started through the API.
- All _local_ operators that are explicitly switched to run remotely with the `remote` modifier.

Not sure on the wording. I think we should start marking the default operator location visibly in the operator docs so that this is easier to understand.

mavam replied:
I tweaked the description. Resolving, since we paired on it.

dominiklohmann commented on web/docs/get-started.md:
```diff
@@ -205,6 +205,14 @@ storage.
 
+:::caution Unsafe Pipelines
+Some operators are inherently unsafe because of their side effects, e.g.,
+reading a file, acquiring packets from the network, or using a third-party
+library. We do not allow these operators by default in managed pipelines. You
```
What do you mean with "or using a third-party library?" The restriction only applies to operators that are marked _local_ internally. For example, we do allow interacting with third-party libraries like Apache Parquet or simdjson in operators that can safely run in any location.

mavam replied:
> What do you mean with "or using a third-party library?"

Something that's not a native operator implementation. An example would be libpcap, although that's also already covered by the previous point. Basically anything that a user can controls with one level of indirection.

> The restriction only applies to operators that are marked local internally. For example, we do allow interacting with third-party libraries like Apache Parquet or simdjson in operators that can safely run in any location.

I don't think the concept of "local" is relevant here. Managed pipelines often run all of their operators on a single node. The user-facing option has `unsafe` in the name, so you need to come from a perspective of security in the argumentation.

mavam commented on web/docs/operators/transformations/flatten.md:
```diff
@@ -10,23 +10,27 @@ flatten [<separator>]
+   the nested records into the parent scope by creating two new fields `x.a` and
+   `x.b`.
+2. **Lists**: merge nested lists into a single (flat) list. For example,
+   `GENERIC-EXAMPLE-MISSING`.
```
@Dakostu or @dominiklohmann, please explain to the user how list flattening works.

Dakostu replied:
Nested lists will be flattened into a singular list, in the same order of the innermost list elements.

Example:
`{"foo": [[[2]],[[3,1]],[[4]]]}` -> `{"foo": [2,3,1,4]}`

For each record inside a list, the operator will flatten its values into a list in a singular record, in the same order of the record's elements. 

`{"foo": [{"a": 2, "b": 1}, {"a": 4}]}` -> `{"foo.a": [2, 4], "foo.b": [1, null]}`

Nested lists inside records will also be flattened.

`{"foo": [{"a": [[2, 23], [1,16]], "b": [1]}, {"a": [[4]]}]}` -> `{"foo.a": [2, 23, 1, 16, 4], "foo.b": [1]}`

mavam replied:
Thanks, @Dakostu. I've added those examples to the docs.

jachris commented on web/docs/setup-guides/deploy-a-node/README.md:
```diff
@@ -0,0 +1,203 @@
+
+## Install a node
+
+Choose between binary package and a Docker-based setup to install a node.
```
```suggestion
Choose between a binary package and a Docker-based setup to install a node.
```

jachris commented on web/docs/setup-guides/use-the-app/README.md:
```diff
@@ -0,0 +1,81 @@
+![Configurator](app-configurator.excalidraw.svg)
+
+Follow INSTRUCTIONS for a concrete sequence of steps to deploy your node. Once
+your node connects, you get a notificatio nand see it in [overview](#overview).
```
```suggestion
your node connects, you get a notification and see it in [overview](#overview).
```

jachris commented on web/docs/user-guides/export-from-a-node.md:
```diff
@@ -5,24 +5,22 @@ sidebar_position: 4
 # Export from a node
 
 Exporting (or *querying*) data is equivalent to [running a
```
Unrelated to your diff, but "equivalent" seems to be the wrong choice here.

```suggestion
Exporting (or *querying*) data can be done by [running a
```

jachris commented on web/docs/user-guides/import-into-a-node.md:
```diff
@@ -5,23 +5,22 @@ sidebar_position: 3
 # Import into a node
 
 Importing (or *ingesting*) data is equivalent to [running a
```
Same here.

jachris commented on web/docs/user-guides/import-into-a-node.md:
```diff
@@ -5,23 +5,22 @@ sidebar_position: 3
-previously. How does it work? The `tenzir` process schedules the first three
-operators in the local process and attempts to place the final `import` operator
-at a remote node, by transparently connecting to the default endpoint
-`127.0.0.1:5158`.
+pipeline successfully, you need to [deploy a
+node](../setup-guides/deploy-a-node/README.md) first. In the Tenzir app, you
+always have a node, but it will fail on the command line without one. The reason
+is that the `tenzir` process schedules the first three operators in the local
+process and attempts to place the final `import` operator at a remote node, by
+transparently connecting to the (default) endpoint `127.0.0.1:5158`.
```
This whole paragraph seems like an implementation detail to me. It would suffice to point out that one needs a node to import data into a node.

jachris commented on web/docs/user-guides/reshape-data/README.md:
```diff
@@ -24,8 +24,8 @@ input with an [expression](../../language/expressions.md).
 
```
Typo above: explaiend

jachris commented on web/docs/user-guides/reshape-data/README.md:
````diff
@@ -24,8 +24,8 @@ input with an [expression](../../language/expressions.md).
-```bash
-tenzir 'export | where #schema == "suricata.alert"'
+```
+export | where #schema == "suricata.alert
````
```suggestion
export | where #schema == "suricata.alert"
```

jachris commented on web/docs/user-guides/reshape-data/README.md:
````diff
@@ -100,8 +100,9 @@ tenzir 'export | where #schema == "suricata.alert"'
+export
+| where 10.10.5.0/25 && (orig_bytes > 1 Mi || duration > 30 min)
 ```
 
````
Not in your diff, but: "The above example extracts connections that either have sent more than 1 MiB or lasted longer than 30 minutes" does not mention the subnet constraint.

jachris commented on web/docs/user-guides/reshape-data/README.md:
```diff
@@ -737,54 +772,234 @@ For example, one way to compute a histogram over the entire persisted dataset is
 </details>
+
+The above pipeline performs a full scan over the data at the node. Tenzir's
+pipeline optimizer pushes down predicate to avoid scans when possible. Consider
```
```suggestion
pipeline optimizer pushes down predicates to avoid scans when possible. Consider
```

jachris commented on web/docs/user-guides/run-a-pipeline/README.md:
```diff
@@ -0,0 +1,118 @@
+On the [command line](../../command-line.md), run `tenzir <pipeline>` where
+`<pipeline>` is the definition of the pipeline as shown in the examples.
+
+The following invariants apply:
+
+1. If you do not provide source, `read json` will be added
+2. If you do not provide sink, `write json` will be added
```
These are not really invariants and this is not complete. For example, we add `save stdout` when the pipeline ends with bytes. But this might be okay, just pointing it out.

jachris commented on web/docs/user-guides/show-available-schemas.md:
```diff
@@ -4,14 +4,20 @@ sidebar_position: 5
 # Show available schemas
 
+:::caution Currently CLI only
+This feature is currently only available on the command line using the `tenzir`
```
```suggestion
This feature is currently only available on the command line using the `tenzir-ctl`
```

jachris commented on web/docs/user-guides/transform-data-at-rest.md:
```diff
@@ -4,6 +4,12 @@ sidebar_position: 6
 # Transform data at rest
 
+:::caution Currently CLI only
+This feature is currently only available on the command line using the `tenzir`
```
```suggestion
This feature is currently only available on the command line using the `tenzir-ctl`
```

(or just drop the `using ...` part)

mavam commented on web/docs/get-started.md:
```diff
@@ -1,296 +1,434 @@
-<TabItem value="debian" label="Debian">
+Let's run a few example [pipelines](language/pipelines.md). Every account comes
+with a pre-installed demo node, but you ultimately want to [bring your
+own](setup-guides/deploy-a-node/README.md). Follow along by copying the below
+examples and pasting them into the [Explorer](https://app.tenzir.com/explorer).
```
A matter of style, ignoring for now.

mavam commented on web/docs/operators/sources/from.md:
````diff
@@ -43,6 +43,7 @@ the individual formats for more information.
 Read bytes from stdin and parse them as JSON.
 
 ```
+read json
````
Removing.
